### PR TITLE
NotificationStore with device registration/unregistration

### PR DESCRIPTION
**Pull request by aforcier**
Adds a basic `NotificationStore`, with support for registering and unregistering the device for push notifications with WordPress.com (using a GCM/FCM token).

Also adds new support for a site ID parameter during registration, which configures the server to only send notifications for that site (and not for any site on the account). This is intended for use with the WooCommerce app.

This replicates much of the registration/deregistration work done in the WordPress app - we could migrate that over to use the `NotificationStore` at some point too.

**Not ready for review**

### TODO:
* [x] Move device registration parameter config to FluxC
* [x] Move UUID creation and storage logic to FluxC
* [x] Define preset values for app key
* [x] Add tests

### CODE REVIEW

File: example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_NotificationTest.kt [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        // No site was specified, so the site selection param should be omitted
+        assertNull(requestBodyMap["selected_blog_id"])
+
+        val payload = lastAction!!.payload as RegisterDeviceResponsePayload
+
+        assertEquals(responseJson.get("ID").asString, payload.deviceId)
+    }
+
+    @Test
+    fun testRegistrationWooCommerce() {
```
Comment (nbradbury) [inline example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_NotificationTest.kt]: Should there also be a test for unregistering the device from push notifs?

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/notifications/NotificationRestClient.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                "device_token" to gcmToken,
+                "device_family" to "android",
+                "app_secret_key" to appKey.value,
+                "device_name" to deviceName,
+                "device_model" to "${Build.MANUFACTURER} ${Build.MODEL}",
+                "app_version" to PackageUtils.getVersionName(appContext),
+                "version_code" to PackageUtils.getVersionCode(appContext).toString(),
+                "os_version" to Build.VERSION.RELEASE,
+                "device_uuid" to uuid,
+                "selected_blog_id" to site?.siteId.toString().takeIf { site != null }
```
Comment (nbradbury) [inline fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/notifications/NotificationRestClient.kt]: I'm confused by this line. If the passed site _is_ null, then we end up setting `selected_blog_id` to "null."

screen shot 2018-09-22 at 10 57 59 am

File: example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_NotificationTest.kt (hunk 1 above)
Comment (aforcier) [reply example/src/androidTest/java/org/wordpress/android/fluxc/mocked/MockedStack_NotificationTest.kt]: Sure! Added one in 4470a01.

File: fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/notifications/NotificationRestClient.kt (hunk 2 above)
Comment (aforcier) [reply fluxc/src/main/java/org/wordpress/android/fluxc/network/rest/wpcom/notifications/NotificationRestClient.kt]: Ah, my mistake - I had a few versions of how to do this and I must've garbled the end result that I actually committed.

Fixed in fe77af6 - that line is supposed to return a null value overall if `site == null`, but then the whole thing needs to be wrapped in a `listOfNotNull()` to filter out any `null` values before creating a map (it's a bit verbose unfortunately).

Ultimately it didn't matter because null body params are filtered out by Volley before being sent, but we may as well do it properly here.

I also updated the corresponding test line to more accurately check that the property is undefined rather than null - this wouldn't have failed the test before anyway though, since the test checks the params actually sent to the server, by which time the `null` was stripped out.

**Comment by nbradbury** [comment]
Looks good! :shipit:
